services/src/main/java/org/keycloak/theme/DefaultThemeManager.java

```diff
@@ -276,6 +277,32 @@ private Map<Locale, Properties> getMessagesByLocale(String baseBundlename, Local
             }
         }
 
+        protected void addlocaleTranslations(Locale locale, Properties m) throws IOException {
+            for (String l : getProperties().getProperty("locales", "").split(",")) {
+                l = l.trim();
+                String key = "locale_" + l;
+                String label = m.getProperty(key);
+                if (label == null) {
+                    String rl = l;
+                    if (l.equals("zh-CN")) {
+                        rl = "zh-HANS";
+                    } else if (l.equals("zh-TW")) {
+                        rl = "zh-HANT";
+                    }
+                    Locale loc = Locale.forLanguageTag(rl);
+                    label = capitalize(loc.getDisplayName(locale), locale);
+                    if (!Objects.equals(loc, locale)) {
+                        label += " (" + capitalize(loc.getDisplayName(loc), loc) + ")";
+                    }
+                    m.put(key, label);
+                }
+            }
+        }
+
+        private String capitalize(String name, Locale locale) {
+            return name.substring(0, 1).toUpperCase(locale) + name.substring(1);
```

Comment:
Not sure if this logic is sound when dealing with different locales, e.g. left-to-right languages, or languages that have Han characters like Kanji.